### PR TITLE
Hide invites in a hidden room

### PR DESCRIPTION
**Pull request by BtbN**
Makes use of version 9 rooms ability to be restricted to allow users who are members of another rooms.
Joins all puppeted users into the hidden room first, if one is set for the channel.
Avoids invite spam in the actual rooms.

Migration of existing rooms is possible via room level command, if they are first migrated to version 9 via external means (most Clients can do that with a simple command).

### CODE REVIEW

File: heisenbridge/__main__.py
Comment (hifi) [inline heisenbridge/__main__.py:622]: I think we should test if the HS supports v9 before even doing this, we can then just test if `hidden_room_id` is set everywhere to know if it exists and is supported.

File: heisenbridge/appservice.py
Comment (hifi) [inline heisenbridge/appservice.py:75]: If we check for the supported room version we don't need to have this code path.

File: heisenbridge/channel_room.py
Comment (hifi) [inline heisenbridge/channel_room.py:143]: I'd call this `UPGRADE` which would do the following:

if room version <9 AND the HS supports it, create new room (min v9), copy all user configurable room data (name, avatar etc.), invite the user, invalidate old room and make it gc'd, doing a real room upgrade doesn't really work with heisenbridge as it does not intentionally keep the rooms that are not actively used and a tombstoned history room wouldn't fit in that category
if room version >=9, just add join rules

I'm also fine if it checks that the room version is <9 and then says you need to leave the room and join the channel again to upgrade manually.

This command can be "slow" in the sense it fetches and checks things as it's not a hot path.

File: heisenbridge/private_room.py
Comment (hifi) [inline heisenbridge/private_room.py:284]: Don't store this in the config since it's a state value, fetch it during bridge initialization and set when creating a room.

**Review by hifi — CHANGES_REQUESTED** [review]
Some comments, in general this is in the right direction!

Additionally since this is a bit experimental it'd be good to have it optional first so that the bridge admin needs to enable it to be the new default. As join rules can be updated on-the-fly it would also be great if it could be disabled/removed as well with the same upgrade command if something goes horribly wrong.

I 100% expect this to be the new default soon after it's been thoroughly tested nothing blows up in heavy use.

Great work!

File: heisenbridge/private_room.py
Comment (BtbN) [reply heisenbridge/private_room.py:284]: It's being used by the join function to determine if the room has a hidden room "attached".
So you mean it should instead fetch the join_rule during init, and read it from there?

File: heisenbridge/channel_room.py
Comment (BtbN) [reply heisenbridge/channel_room.py:143]: I'm not a huge fan of performing that kind of complex upgrade in the bridge. If a user wants it, most clients can do it.
It's also apparently not easily possible to check if a room version is >= something, since the version is not neccesarily numeric.
So basically, it'd have to be a check if the room value is not any of versions 1 to 8, which is a bit ugly.

So currently the approach is to try and add the join_rule, and if it fails, tell the user to upgrade/re-join the room. Though I'm not sure if adding it will really fail?

**Comment by BtbN** [comment]
Should it be like a command line option, or what kind of opt-in did you think about?

File: heisenbridge/channel_room.py
Comment (hifi) [reply heisenbridge/channel_room.py:143]: Clients can't do it as they don't have the required PL. It's also something that wouldn't work without bridge support since upgrading always creates a new room _anyway_ so the bridge needs to move its operations there.

Just check if the room is of unsupported version `in [1,2,3,4,5,6,7,8]` or something and then tell them to leave it if you don't want to try adding support for real upgrades.

File: heisenbridge/private_room.py
Comment (hifi) [reply heisenbridge/private_room.py]: Yes, fetch the join_rule during init and fill that value from there instead of storing it twice.

**Comment by hifi** [comment]
An admin command to enable the new behavior would be good. We can make it the default a bit later when it's been battle tested.

**Comment by BtbN** [comment]
This should cover most things.
Mind you, I have not yet deployed and tested this code, so I would not recommend merging yet, if no further comments arise.

File: heisenbridge/appservice.py
Comment (hifi) [inline heisenbridge/appservice.py:51]: No need to test this again, if `hidden_room_id` is set we already know v9 is supported.

File: heisenbridge/__main__.py
Comment (hifi) [inline heisenbridge/__main__.py:610]: If this was a real room type like in the original revision you could make it a validation rule that if the configuration option is not enabled the room is invalidated and it goes into cleanup queue, then you wouldn't need this double variable hack. I'm not blocking merging because of it but it would be slightly cleaner to keep it as a room type so it can be collected for cleanup when the feature is disabled.

File: heisenbridge/__main__.py
Comment (hifi) [inline heisenbridge/__main__.py:416]: This must go through all rooms and remove the join rule I'm afraid.

**Review by hifi — CHANGES_REQUESTED** [review]
Looks pretty good now, few comments.

File: heisenbridge/appservice.py
Comment (BtbN) [reply heisenbridge/appservice.py:51]: It's not about testing if it's supported, it's about querying the default version.
v9 can be supported but not the default. The whole idea here is to not hold rooms back from getting a higher version, so I only explicitly request version 9 if 1..8 is the default.
I thought about saving this information from the initial query, but that would cause some confusion if the default is changed at runtime of the bridge.

File: heisenbridge/__main__.py
Comment (BtbN) [reply heisenbridge/__main__.py:610]: I initially had it as proper room type, so it'll definitely be nicer. I'll see if it can be easily revived.

File: heisenbridge/__main__.py
Comment (BtbN) [inline heisenbridge/__main__.py:604]: This section will be dropped before merging, it only exists to migrate my own server.

File: heisenbridge/__main__.py
Comment (hifi) [inline heisenbridge/__main__.py:403]: Shouldn't this return `False` instead?

Comment (hifi) [reply heisenbridge/__main__.py:403]: Ah, no, nevermind.

**Comment by hifi** [comment]
Okay, we'll continue refining this in the tree when needed, I'm happy enough with the current version to merge, thanks!

**Pull request merged**
